File: solutions/javascript/src/algorithm/multipledictionary/Exercise3.js
import DynamicMultipleDictionaryADT from "../../structure/dynamic/DynamicMultipleDictionaryADT.js";


function filtrarPorRango(diccionario, minimo, maximo) {
  const resultado = new DynamicMultipleDictionaryADT();
  const claves = diccionario.getKeys();

  for (let i = 0; i < claves.length; i++) {
    const clave = claves[i];

    try {
      const valores = diccionario.get(clave);
      for (let j = 0; j < valores.length; j++) {
        const valor = valores[j];
        if (valor >= minimo && valor <= maximo) {
          resultado.add(clave, valor);
        }
      }
    } catch (_) {
      // ignorar si la clave no existe (no debería ocurrir)
    }
  }

  return resultado;
}


const diccionario = new DynamicMultipleDictionaryADT();
diccionario.add(1, 10);
diccionario.add(1, 20);
diccionario.add(2, 5);
diccionario.add(2, 30);
diccionario.add(3, 100);

const filtrado = filtrarPorRango(diccionario, 10, 30);
const clavesFiltradas = filtrado.getKeys();

for (let i = 0; i < clavesFiltradas.length; i++) {
  const clave = clavesFiltradas[i];
  const valores = filtrado.get(clave);
  console.log(`Clave ${clave}: ${valores.join(', ')}`);
}

File: solutions/python/algorithm/multipledictionary/Exercise3.py
from structure.dynamic.DynamicMultipleDictionaryADT import DynamicMultipleDictionaryADT


def filtrar_por_rango(diccionario: DynamicMultipleDictionaryADT, minimo: int, maximo: int) -> DynamicMultipleDictionaryADT:
    resultado = DynamicMultipleDictionaryADT()
    claves = diccionario.get_keys().get_all()

    for clave in claves:
        try:
            valores = diccionario.get(clave)
            for valor in valores:
                if minimo <= valor <= maximo:
                    resultado.add(clave, valor)
        except KeyError:
            pass  # ignorar claves que no existen (aunque no debería pasar)

    return resultado


d = DynamicMultipleDictionaryADT()
d.add(1, 10)
d.add(1, 20)
d.add(1, 5)
d.add(2, 100)
d.add(2, 25)
d.add(3, 30)

filtrado = filtrar_por_rango(d, 10, 30)
filtrado.mostrar()

